src/sentry/region_to_control/consumer.py

```diff
@@ -70,25 +71,49 @@ def flush_batch(self, batch: Sequence[RegionToControlMessage]):
     def _flush_batch(self, batch: Sequence[RegionToControlMessage]):
         for row in batch:
             if row.user_ip_event:
-                updated, created = UserIP.objects.create_or_update(
-                    values=dict(
-                        user_id=row.user_ip_event.user_id,
-                        ip_address=row.user_ip_event.ip_address,
-                        last_seen=row.user_ip_event.last_seen,
-                        country_code=row.user_ip_event.country_code,
-                        region_code=row.user_ip_event.region_code,
-                    )
-                )
-
-                if created:
-                    metrics.incr("region_to_control.consumer.user_ip_event.created")
-                elif updated:
-                    metrics.incr("region_to_control.consumer.user_ip_event.updated", amount=updated)
-                else:
-                    # This happens when there is an integrity error adding the UserIP -- such as when user is deleted,
-                    # or the ip address does not match the db validation.  This is expected and not an error condition
-                    # in low quantities.
-                    metrics.incr("region_to_control.consumer.user_ip_event.stale_event")
+                self._handle_user_ip_event(row.user_ip_event)
+            if row.audit_log_event:
+                self._handle_audit_log_event(row.audit_log_event)
+
+    def _handle_audit_log_event(self, audit_log_entry: AuditLogEvent, reentry=False):
```

Comment:
@corps is there a chance that these handlers be co-located with where the related Models are defined?

Reply:
As an unobvious as it is, I feel the intentional separation is good.  Firstly, the logic of serialization and deserialization is, in fact, co-located in the database model.  So that makes sense.  However, 'handling' logic here is not necessarily limited to just model concerns; for instance, this method currently also handles replication delay issues wrt to foreign keys as well, something I think ought not to be hugged too closely to the model in reality.  


In any case, me and ryan are moving close to having 'stable' silo tests running on master that are going to help lock down and make visible changes that run counter to these new code paths.